File: 278. First Bad Version.cs
/* The isBadVersion API is defined in the parent class VersionControl.
      bool IsBadVersion(int version); */

public class Solution : VersionControl {
    public int FirstBadVersion(int n) {
        int  l = 1, h = n, p = (n-1)/2 + 1;
        while(l<h){
            if(IsBadVersion(p)){
                h = p;
            }
            else{
                l = p+1;
            }
            p = l/2+h/2;
        }
        return h;
    }
}

File: 278. First Bad Version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int a = 2, b = n, mid;
        while (a <= b) {
            mid = a + (b - a) / 2;
            if (isBadVersion(mid) == true){
                if(isBadVersion(mid - 1) == false) 
                    return mid;
                else
                    b = mid - 1;
            }
            else
                a = mid + 1;
        }
        return b;
    }
}
